openssl-sys/build.rs

```diff
@@ -1,86 +1,323 @@
-                }
-                return;
-            }
+    if !Path::new(&include_dir).exists() {
+        panic!("OpenSSL include directory does not exist: {}",
+               include_dir.to_string_lossy());
+    }
+
+    println!("cargo:rustc-link-search=native={}", lib_dir.to_string_lossy());
+    println!("cargo:include={}", include_dir.to_string_lossy());
+
+    let version = validate_headers(&[include_dir.clone().into()],
+                                   &[lib_dir.clone().into()]);
+
+    let libs = if (version.contains("OpenSSL 1.0.1") ||
+                   version.contains("OpenSSL 1.0.2")) &&
+                  target.contains("windows") {
+        ["ssleay32", "libeay32"]
+    } else if target.contains("windows") {
+        ["libssl", "libcrypto"]
+    } else {
+        ["ssl", "crypto"]
+    };
+
+    let kind = determine_mode(Path::new(&lib_dir), &libs);
+    for lib in libs.iter() {
+        println!("cargo:rustc-link-lib={}={}", kind, lib);
+    }
+}
+
+fn find_openssl_dir(target: &str) -> OsString {
+    let host = env::var("HOST").unwrap();
+
+    if host.contains("apple-darwin") && target.contains("apple-darwin") {
+        let homebrew = Path::new("/usr/local/opt/openssl");
+        if homebrew.exists() {
+            return homebrew.to_path_buf().into()
         }
-        if let Some(mingw_paths) = get_mingw_in_path() {
-            for path in mingw_paths {
-                println!("cargo:rustc-link-search=native={}", path);
-            }
+        let homebrew = Path::new("/usr/local/opt/openssl@1.1");
+        if homebrew.exists() {
+            return homebrew.to_path_buf().into()
+        }
+
+        let system = Path::new("/usr/lib/libssl.0.9.8.dylib");
+        if system.exists() {
+            panic!("
+
+Failed to find an approprate OpenSSL installation. It looks like you're
+compiling on macOS, where the system contains a version of OpenSSL 0.9.8. This
+crate no longer supports OpenSSL 0.9.8.
+
+As a consumer of this crate, you can fix this error by using Homebrew to
+install the `openssl` package, or as a maintainer you can use the openssl-sys
+0.7 crate for support with OpenSSL 0.9.8.
+
+Unfortunately though the compile cannot continue, so aborting.
+
+");
         }
     }
 
-    let libs_env = env::var("OPENSSL_LIBS").ok();
-    let libs = match libs_env {
-        Some(ref v) => v.split(":").collect(),
-        None => if target.contains("windows") {
-            if get_mingw_in_path().is_some() && lib_dir.is_none() && include_dir.is_none() {
-                vec!["ssleay32", "eay32"]
-            } else {
-                vec!["ssl32", "eay32"]
-            }
-        } else {
-            vec!["ssl", "crypto"]
+    try_pkg_config();
+
+    // TODO: try to be more helpful in this message, e.g.:
+    //
+    // * if we're on linux, recommend apt-get something or other
+    // * if we're on Windows, link to a README explaining more
```

Comment:
I'd love to handle this specifically (compiling on Windows) before landing. Do you perhaps have suggestions about what to put in this error though?